docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+Custom Attestation Data Formats for Open Enclave V3 Updates
+================
+
+As custom attestation plugins are developed and features are requested, new APIs were added to:
+
+1. Query built-in SGX attesters and verifiers to be registered.
+2. Select a format of the evidence that the registered attesters can generate.
+3. Query evidence parameters to be used as optional inputs when generating evidences.
+
+The following API was affected:
+
+1. Built-in SGX attesters and verifiers
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
+   for (n = 0; n < attesters_length; n++)
+      oe_register_attester(&attesters[n], NULL, 0);
```

Comment:
another option would be to to provide an api `oe_register_attesters` which takes an array of UUIDs

Reply:
The plugin registration was designed this way in V1 because early inputs asked for the requirements for application to pass in configuration data (the 2nd parameter) during plugin registration.

Considering the fact that some TEEs don't know available attesters/verifiers until run time (e.g. checking installation, back-end setup, etc), we are adding extra interface to get attesters/verifiers.

---
In reply to: [393247888](https://github.com/openenclave/openenclave/pull/2621#discussion_r393247888) [](ancestors = 393247888)

Reply:
I tend to agree that we can have a single function like `oe_register_attesters()`, or `oe_initialize_attesters()` that enumerate all attester plugins and registers them.

I see that the existing `oe_register_attester()` takes an optional `config_data` parameters. But for SGX, I don't see need of any optional plugin-in specific, configuration data. To ensure TEE agnostic, the application software should not have to deal with TEE-specific configuration anyways.

Reply:
Before the plugin is registered, OE SDK does not know how to map UUID to plugin.

---
In reply to: [393381921](https://github.com/openenclave/openenclave/pull/2621#discussion_r393381921) [](ancestors = 393381921)

Reply:
config_data in oe_register_attester() was an early requirement when the plugin API was first designed. It'd be a breaking change if we disallow it now.

---
In reply to: [393873283](https://github.com/openenclave/openenclave/pull/2621#discussion_r393873283) [](ancestors = 393873283)

Reply:
If there was an answer to my question, I didn't understand it... can someone reword an answer if so?

Reply:
If we "either (a) having all registered by default, or (b) having them be in metadata like the .conf file", the developers lose the flexibility to customize each plugin instance with "config_data" like this: oe_register_attester(plugin, config_data, config_data_size). 

One example is that attester can generate different result based on policy that's generated at run time.

---
In reply to: [394651138](https://github.com/openenclave/openenclave/pull/2621#discussion_r394651138) [](ancestors = 394651138)

Reply:
I don't see any justification in this doc for having "configuration data" required at registration time.  Above said "early inputs asked for the requirements for application to pass in configuration data (the 2nd parameter) during plugin registration" but I see no pointers to any public input of that sort, nor any motivation in this document.  Either remove the capability, or provide the motivation in this document.

For example,
1) what is such "configuration data" needed for? I.e., what can you do with it that someone needs, that you can't do without it?
2) why would it be needed at registration time instead of post registration?
3) is config data required for all apps on all platforms, as just for advanced apps on particular platforms?  (which would motivate having the automatic registration via (a) or (b) as the recommended flow)

We need to make sure we understand what problem is trying to be solved before we can agree on whether this is the right way to solve it.

Reply:
> Given the time constraint, I think sticking to the design in the previous release seems logical. 

Can you say more what the "time constraint" is?  It sounds like you're proposing inserting another version of an experimental API that we might likely immediately deprecate in the next PR, which means that products should not really use it yet.   Why not get the API right (or at least as best as we know now) without hurrying?

Reply:
The "configuration data" was part of the original plugin design to allow an plug-in user to pass in instance data to an plugin instance. For example, this config data could contain policy information that a plug-in needs for its validation. We decided to do it during the registration time to reduce the number of calls a client needs to make.

Reply:
"We decided to do it during the registration time to reduce the number of calls a client needs to make." makes no sense, since (a) or (b) would reduce the number of calls a client would make.  And I'm not sure what "We decided" means.  Design decisions are made here in the open, so if there was a decision documented here recently, please point me to it, since I cannot find any such decision.